README.md
[![Build Status](https://travis-ci.org/hitode909/APISchema.svg?branch=master)](https://travis-ci.org/hitode909/APISchema)
# NAME

APISchema - Schema for API

# SYNOPSIS

    # bmi.def

    resource figure => {
        type => 'object',
        description => 'Figure, which includes weight and height',
        properties => {
            weight  => {
                type => 'number',
                description => 'Weight(kg)',
                example => 50,
            },
            height  => {
                type => 'number',
                description => 'Height(m)',
                example => 1.6,
            },
        },
        required => ['weight', 'height'],
    };

    resource bmi => {
        type => 'object',
        description => 'Body mass index',
        properties => {
            value  => {
                type => 'number',
                description => 'bmi value',
                example => 19.5,
            },
        },
        required => ['value'],
    };

    POST '/bmi/' => {
        title           => 'BMI API',
        description     => 'This API calculates your BMI.',
        destination     => {
            controller  => 'BMI',
            action      => 'calculate',
        },
        request         => 'figure',
        response        => 'bmi',
    };

    # main.pl

    use APISchema::DSL;
    my $schema = APISchema::DSL::process {
        include 'bmi.def';
    };

    # Routing
    use APISchema::Generator::Router::Simple;
    my $router = do {
        my $generator = APISchema::Generator::Router::Simple->new;
        $generator->generate_router($schema);
    };

    # Inject routes to an existing router object
    my $router = Router::Simple->new;
    $router->connect(...);
    my $generator = APISchema::Generator::Router::Simple->new;
    $generator->inject_routes($schema => $router);

    # Documentation
    use APISchema::Generator::Markdown;
    print do {
        my $generator = APISchema::Generator::Markdown->new;
        $generator->format_schema($schema);
    };

    # Middleware (in app.psgi)

    enable "APISchema::ResponseValidator", schema => $schema;
    enable "APISchema::RequestValidator", schema => $schema;

# DESCRIPTION

APISchema is Schema for API

It provides DSL to describe API specification schema.
It generates router, validator, document from API schema.

# LICENSE

Copyright (C) hitode909 and tarao.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHORS

hitode909 &lt;hitode909@gmail.com>

tarao &lt;tarao.gnn@gmail.com>


Changes
Revision history for Perl extension APISchema

{{$NEXT}}

1.37 2018-02-08T10:31:50Z
  - Allow `example: undef` and serve document

1.36 2017-12-25T12:45:24Z
  - show enum values in type column

1.35 2017-12-13T14:30:28Z
  - MockServer serves default status codes in definitions

1.34 2017-11-30T16:29:26Z
  - Define prototype of type before calling it

1.33 2017-11-30T14:39:09Z
  - Resolve anyOf, allOf and oneOf keywords when serving Mock Server and API Document

1.32 2017-11-30T06:08:51Z
  - MockServer always returns bytes

1.31 2017-11-29T02:30:47Z
  - MockServer support `response => {body => 'bmi', encoding => 'json'}` syntax

1.30 2017-11-20T17:00:38Z
   - Empty input could pass validations unexpectedly
   - Add . to lib

1.29 2017-11-20T01:34:11Z
   - Handle boolean on generating document(1.28 was broken)

1.28 2017-11-17T09:49:53Z
   - Handle boolean on generating document

1.27 2017-09-19T09:58:40Z
   - Sort keys when encoding JSON

1.26 2017-08-30T02:20:58Z
   - Expand `FETCH` method to `GET` and `HEAD`

1.25 2017-08-29T02:31:53Z
   - You can inherit Plack::App::APISchema::Document and customize template

1.24 2016-06-10T10:01:57Z
   - Fix document

1.23 2016-06-10T09:52:33Z
   - Inject routes to an existing router object

1.22 2016-04-20T06:08:32Z
   - Update Valiemon

1.21 2015-11-24T02:35:14Z
    - Added missing prereqs as reported by CPANTS

1.2 2015-11-20T04:37:28Z

    - X-Error-Cause header indicates whether the response comes from app or validator

1.1 2015-11-19T06:32:07Z
    - Add method and path info to documentation TOC (Sixeight++)

1.04 2015-10-15T10:38:54Z
    - mock server feature

1.03 2015-10-06T07:25:49Z
    - Determine the status code from the detail of failure (aereal++)

1.02 2015-10-06T07:12:24Z
    - Prettify the output of examples (aereal++)

1.01 2015-09-04T09:32:37Z
    - Recognize HTTP PATCH method (aereal++)

1.00 2015-08-27T10:48:03Z
    - Support schema which contains unicode characters
    - Enable `use utf8` in DSL file

0.03 2015-08-18T02:06:45Z
    - Validation error has more details. expected: expected schema, actual: actual input data

0.02 2015-03-16T02:35:36Z
    - Add a status_code parameter to P::MW::AS::RequestValidator

0.01 2015-02-18T10:41:05Z

    - original version


lib/APISchema.pm
package APISchema;
use 5.014;
use strict;
use warnings;

our $VERSION = "1.33";

1;
__END__

=encoding utf-8

=head1 NAME

APISchema - Schema for API

=head1 SYNOPSIS

    # bmi.def

    resource figure => {
        type => 'object',
        description => 'Figure, which includes weight and height',
        properties => {
            weight  => {
                type => 'number',
                description => 'Weight(kg)',
                example => 50,
            },
            height  => {
                type => 'number',
                description => 'Height(m)',
                example => 1.6,
            },
        },
        required => ['weight', 'height'],
    };

    resource bmi => {
        type => 'object',
        description => 'Body mass index',
        properties => {
            value  => {
                type => 'number',
                description => 'bmi value',
                example => 19.5,
            },
        },
        required => ['value'],
    };

    POST '/bmi/' => {
        title           => 'BMI API',
        description     => 'This API calculates your BMI.',
        destination     => {
            controller  => 'BMI',
            action      => 'calculate',
        },
        request         => 'figure',
        response        => 'bmi',
    };

    # main.pl

    use APISchema::DSL;
    my $schema = APISchema::DSL::process {
        include 'bmi.def';
    };

    # Routing
    use APISchema::Generator::Router::Simple;
    my $router = do {
        my $generator = APISchema::Generator::Router::Simple->new;
        $generator->generate_router($schema);
    };

    # Inject routes to an existing router object
    my $router = Router::Simple->new;
    $router->connect(...);
    my $generator = APISchema::Generator::Router::Simple->new;
    $generator->inject_routes($schema => $router);

    # Documentation
    use APISchema::Generator::Markdown;
    print do {
        my $generator = APISchema::Generator::Markdown->new;
        $generator->format_schema($schema);
    };

    # Middleware (in app.psgi)

    enable "APISchema::ResponseValidator", schema => $schema;
    enable "APISchema::RequestValidator", schema => $schema;

=head1 DESCRIPTION

APISchema is Schema for API

It provides DSL to describe API specification schema.
It generates router, validator, document from API schema.

=head1 LICENSE

Copyright (C) hitode909 and tarao.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHORS

hitode909 E<lt>hitode909@gmail.comE<gt>

tarao E<lt>tarao.gnn@gmail.comE<gt>

=cut

